File: internal/repository/member_repository.go
// Package repository contains database access code.
package repository

import (
	"context"
	"database/sql"
	"fmt"
	"strings"

	"github.com/Homeria/baeum-maru/internal/domain"
)

type MemberRepository struct {
	db *sql.DB
}

type CreateMemberParams struct {
	MemberNo   string
	Name       string
	GenderCode string
	BirthDate  string
	Phone      string
	Note       string
}

type UpdateMemberParams struct {
	ID         int64
	MemberNo   string
	Name       string
	GenderCode string
	BirthDate  string
	Phone      string
	Note       string
}

func NewMemberRepository(db *sql.DB) *MemberRepository {
	return &MemberRepository{db: db}
}

func (r *MemberRepository) Create(ctx context.Context, params CreateMemberParams) (domain.Member, error) {
	result, err := r.db.ExecContext(ctx, `
INSERT INTO members (member_no, name, gender_code, birth_date, phone, note)
VALUES (NULLIF(?, ''), ?, NULLIF(?, ''), NULLIF(?, ''), NULLIF(?, ''), NULLIF(?, ''));
`, strings.TrimSpace(params.MemberNo), strings.TrimSpace(params.Name), strings.TrimSpace(params.GenderCode), strings.TrimSpace(params.BirthDate), strings.TrimSpace(params.Phone), strings.TrimSpace(params.Note))
	if err != nil {
		return domain.Member{}, fmt.Errorf("insert member: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return domain.Member{}, fmt.Errorf("read member id: %w", err)
	}
	return r.Get(ctx, id)
}

func (r *MemberRepository) Update(ctx context.Context, params UpdateMemberParams) (domain.Member, error) {
	result, err := r.db.ExecContext(ctx, `
UPDATE members
SET member_no = NULLIF(?, ''),
    name = ?,
    gender_code = NULLIF(?, ''),
    birth_date = NULLIF(?, ''),
    phone = NULLIF(?, ''),
    note = NULLIF(?, ''),
    updated_at = CURRENT_TIMESTAMP
WHERE id = ?;
`, strings.TrimSpace(params.MemberNo), strings.TrimSpace(params.Name), strings.TrimSpace(params.GenderCode), strings.TrimSpace(params.BirthDate), strings.TrimSpace(params.Phone), strings.TrimSpace(params.Note), params.ID)
	if err != nil {
		return domain.Member{}, fmt.Errorf("update member %d: %w", params.ID, err)
	}

	affected, err := result.RowsAffected()
	if err != nil {
		return domain.Member{}, fmt.Errorf("read affected member rows: %w", err)
	}
	if affected == 0 {
		return domain.Member{}, sql.ErrNoRows
	}
	return r.Get(ctx, params.ID)
}

func (r *MemberRepository) Get(ctx context.Context, id int64) (domain.Member, error) {
	row := r.db.QueryRowContext(ctx, `
SELECT id, COALESCE(member_no, ''), name, COALESCE(gender_code, ''), COALESCE(birth_date, ''),
       COALESCE(phone, ''), COALESCE(note, ''), created_at, updated_at
FROM members
WHERE id = ?;
`, id)
	return scanMember(row)
}

func (r *MemberRepository) Search(ctx context.Context, query string, limit int) ([]domain.Member, error) {
	if limit <= 0 {
		limit = 50
	}
	if limit > 10000 {
		limit = 10000
	}

	query = strings.TrimSpace(query)
	like := "%" + query + "%"
	rows, err := r.db.QueryContext(ctx, `
SELECT id, COALESCE(member_no, ''), name, COALESCE(gender_code, ''), COALESCE(birth_date, ''),
       COALESCE(phone, ''), COALESCE(note, ''), created_at, updated_at
FROM members
WHERE ? = ''
   OR name LIKE ?
   OR COALESCE(member_no, '') LIKE ?
   OR COALESCE(phone, '') LIKE ?
ORDER BY name, id
LIMIT ?;
`, query, like, like, like, limit)
	if err != nil {
		return nil, fmt.Errorf("search members: %w", err)
	}
	defer rows.Close()

	var members []domain.Member
	for rows.Next() {
		member, err := scanMember(rows)
		if err != nil {
			return nil, err
		}
		members = append(members, member)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate members: %w", err)
	}
	return members, nil
}

type memberScanner interface {
	Scan(dest ...any) error
}

func scanMember(scanner memberScanner) (domain.Member, error) {
	var member domain.Member
	if err := scanner.Scan(
		&member.ID,
		&member.MemberNo,
		&member.Name,
		&member.GenderCode,
		&member.BirthDate,
		&member.Phone,
		&member.Note,
		&member.CreatedAt,
		&member.UpdatedAt,
	); err != nil {
		return domain.Member{}, fmt.Errorf("scan member: %w", err)
	}
	return member, nil
}

File: backend/app/repositories/member_repository.py
"""회원 데이터를 다루는 repository."""

import sqlite3
from typing import Any

from app.core.exceptions import ConflictError
from app.db.database import get_db_connection

# CRUD - Read


def list_members(*, q: str | None = None, include_inactive: bool = False) -> list[dict[str, Any]]:
    query = "SELECT * FROM members"
    conditions: list[str] = []
    params: list[Any] = []
    if q:
        conditions.append("(name LIKE ? OR phone LIKE ? OR member_no LIKE ?)")
        like = f"%{q}%"
        params.extend([like, like, like])
    if not include_inactive:
        conditions.append("is_active = 1")
    if conditions:
        query += " WHERE " + " AND ".join(conditions)
    query += " ORDER BY name"
    with get_db_connection() as conn:
        rows = conn.execute(query, params).fetchall()
    return [dict(row) for row in rows]


def get_member(member_id: int) -> dict[str, Any] | None:
    with get_db_connection() as conn:
        row = conn.execute("SELECT * FROM members WHERE id = ?", (member_id,)).fetchone()
    return dict(row) if row is not None else None


# CRUD - Create/Update/Delete


def create_member(*, member_no: str, name: str, gender: str, phone: str) -> dict[str, Any]:
    with get_db_connection() as conn:
        try:
            cursor = conn.execute(
                "INSERT INTO members (member_no, name, gender, phone) VALUES (?, ?, ?, ?)",
                (member_no, name, gender, phone),
            )
            conn.commit()
        except sqlite3.IntegrityError as error:
            raise ConflictError("member_no_exists", "이미 등록된 회원번호입니다.") from error
        row = conn.execute("SELECT * FROM members WHERE id = ?", (cursor.lastrowid,)).fetchone()
    return dict(row)


def update_member(
    member_id: int, *, name: str, gender: str, phone: str, is_active: bool
) -> dict[str, Any] | None:
    with get_db_connection() as conn:
        cursor = conn.execute(
            """
            UPDATE members
            SET name = ?, gender = ?, phone = ?, is_active = ?, updated_at = CURRENT_TIMESTAMP
            WHERE id = ?
            """,
            (name, gender, phone, int(is_active), member_id),
        )
        conn.commit()
        if cursor.rowcount == 0:
            return None
        row = conn.execute("SELECT * FROM members WHERE id = ?", (member_id,)).fetchone()
    return dict(row)


def delete_member(member_id: int) -> bool:
    with get_db_connection() as conn:
        try:
            cursor = conn.execute("DELETE FROM members WHERE id = ?", (member_id,))
            conn.commit()
        except sqlite3.IntegrityError as error:
            raise ConflictError(
                "member_in_use", "신청 이력이 있어 삭제할 수 없습니다. 비활성화하세요."
            ) from error
    return cursor.rowcount > 0
